prov/gni/src/gnix_mr.c

```diff
@@ -212,20 +212,9 @@ DIRECT_FN int gnix_mr_reg(struct fid *fid, const void *buf, size_t len,
 			1 << GNIX_MR_PAGE_SHIFT);
 
 	/* call cache register op to retrieve the right entry */
-	fastlock_acquire(&domain->mr_cache_lock);
-	if (unlikely(!domain->mr_cache)) {
-		rc = _gnix_mr_cache_init(&domain->mr_cache,
-				&domain->mr_cache_attr);
-		if (rc != FI_SUCCESS) {
-			fastlock_release(&domain->mr_cache_lock);
-			goto err;
-		}
-	}
-
-	rc = _gnix_mr_cache_register(domain->mr_cache,
```

Comment:
This seems a bit dubious. A user shouldn't call into a structure if it hasn't been initialized and now we've migrated the initialization point into the operation. Additionally, this might interfere with close behavior, but I'm not sure at this point. 

I'd advise against taking external locks in a provider agnostic implementation such as we are trying to migrate to.